File: packages/vscode/src/tui-bridge.ts
import {
  type BridgeHandleFn,
  type BridgeServer,
  listenBridge,
  socketAddress,
} from "@groks-beard/mcp"

export const TUI_BRIDGE_STATE_KEY = "groksBeard.tuiBridge.enabled"

export type TuiBridgeDeps = {
  readonly getEnabled: () => boolean
  readonly workspace: () => string | undefined
  readonly handle: BridgeHandleFn
  readonly log: (message: string) => void
}

export class TuiBridge {
  private server: BridgeServer | undefined

  constructor(private readonly deps: TuiBridgeDeps) {}

  get listening(): boolean {
    return this.server?.listening === true
  }

  address(): string | undefined {
    const workspace = this.deps.workspace()
    if (workspace === undefined || workspace === "") return undefined
    return socketAddress({ workspace })
  }

  async sync(): Promise<void> {
    const workspace = this.deps.workspace()
    if (!this.deps.getEnabled() || workspace === undefined || workspace === "") {
      await this.unbind()
      return
    }
    if (this.server !== undefined) return
    const address = socketAddress({ workspace })
    this.server = await listenBridge(address, this.deps.handle)
    this.deps.log(`TUI bridge listening on ${address}`)
  }

  async unbind(): Promise<void> {
    const server = this.server
    this.server = undefined
    if (server !== undefined) await server.close()
  }
}

File: host/src/main/scala/groksbeard/host/TuiBridge.scala
package groksbeard.host

import groksbeard.core.*
import zio.json.*

import scala.scalajs.js

final class TuiBridge(host: McpToolHost, log: String => Unit):
  private var server: Option[NodeServer] = None

  def listening: Boolean = server.isDefined

  def address(workspace: String): String =
    val win     = nodeProcess.platform == "win32"
    val runtime =
      nodeProcess.env.get("XDG_RUNTIME_DIR").flatMap(v => v.toOption).filter(_.nonEmpty).getOrElse(nodeOs.tmpdir())
    SocketAddress.address(workspace, win, TuiBridge.sha256Hex, runtime)

  def sync(enabled: Boolean, workspace: Option[String]): Unit =
    if !enabled || workspace.isEmpty then unbind()
    else if server.isEmpty then bind(workspace.get)

  def unbind(): Unit =
    server.foreach { s =>
      val _ = s.close()
    }
    server = None

  private def bind(workspace: String): Unit =
    val addr = address(workspace)
    if !addr.startsWith("\\\\.\\pipe\\") then
      nodeFs.mkdirSync(nodePath.dirname(addr), js.Dynamic.literal(recursive = true, mode = SocketAddress.SocketDirMode))
      try nodeFs.unlinkSync(addr)
      catch case _: Throwable => ()
    val srv = nodeNet.createServer { (socket: NodeSocket) =>
      var buffer = ""
      socket.setEncoding("utf8")
      socket.on(
        "data",
        (chunk: js.Any) =>
          val (lines, rest) = Ndjson.split(buffer, "" + chunk)
          buffer = rest
          lines.foreach { line =>
            val reply =
              line.fromJson[BridgeRequest] match
                case Left(_)    => McpBridge.fail("", "invalid json")
                case Right(req) =>
                  McpTools.dispatch(req.tool, req.args, host) match
                    case Left(err)  => McpBridge.fail(req.id, err)
                    case Right(res) => McpBridge.ok(req.id, res)
            val _ = socket.write(Ndjson.encode(reply.toJson))
            socket.end()
          },
      )
      socket.on("error", (_: js.Any) => ())
    }
    srv.on("error", (err: js.Any) => log("TUI bridge error: " + err))
    srv.listen(
      addr,
      () =>
        try nodeFs.chmodSync(addr, SocketAddress.SocketMode)
        catch case _: Throwable => ()
        log(s"TUI bridge listening on $addr"),
    )
    server = Some(srv)
  end bind
end TuiBridge

object TuiBridge:
  val StateKey = "groksBeard.tuiBridge.enabled"

  def sha256Hex(text: String): String =
    nodeCrypto.createHash("sha256").update(text).digest("hex")
